src/Support/TypeUtilities.cpp

```diff
@@ -51,6 +51,7 @@ int64_t getRank(Type ty) {
 
 /// Get the number of elements.
 int64_t getNumberOfElements(Type ty) {
+  assert(ty.cast<ShapedType>().hasStaticShape() && "Has unknown dimensions");
```

Comment:
fwiw, I just learned that `x.cast<T>()` is officially deprecated in favor of `cast<T>(x)` https://mlir.llvm.org/deprecation/

Reply:
Thanks! It's good to know. We should have a PR to update our whole code.

Reply:
actually, if we're going to cast ty to ShapedType anyway, how about we change the function to take ShapedType input argument:
```
int64_t getNumberOfElements(ShapedType ty) {
  assert(ty.hasStaticShape() && ..);
  return ty.getNumberOfElements();
}
```

since ShapedType is a type interface you can pass any Type and it will automatically do the cast to ShapedType

Reply:
It's a good idea. Let me change it to `ShapedType`.